core/src/docs/rfcs/2133_append_api.md

```diff
@@ -0,0 +1,106 @@
+- Proposal Name: `append_api`
+- Start Date: 2023-04-26
+- RFC PR: [apache/incubator-opendal#2133](https://github.com/apache/incubator-opendal/pull/2133)
+- Tracking Issue: 
+
+# Summary
+
+Introduce append operations for OpenDAL which allow users to add data to a file. And write operation will be not reentrant.
+
+# Motivation
+
+OpenDAL has the write operation used to create a file and append data to it. This is implemented based on multipart API. However, current approach has some limitations:
+
+- Data could be lost and not readable before w.close() returned Ok(())
+- File can't be appended again after w.close() returned Ok(())
+
+To address these issues, I propose adding an append operation and modifying the write operation to verify consistency between content length and data size. The revised write operation will not be reentrant; however, users can create an appender that provides a reentrant append operation.
+
+# Guide-level explanation
+
+The files created by the write operation can not be modified, only overwritten.
+
+```rust
+async fn write_test(op: Operation) -> Result<()> {
+    let bs = read_from_file();
+    op.write("path_to_file", bs).await?;
+  
+    // The following code will report an error
+    op.write("path_to_file", bs).await?;
+  
+    // Unless specified can overwrite
+    let option = OpWrite::default().with_allow_overwrite(true);
+    // content length must be the same as the data size
+    option.with_content_length(bs.len() as u64);
+    op.write_with("path_to_file", option, bs).await?;
+}
+```
+
+The files created by the append operation can be appended via multipart API or append API.
```

Comment:
multipart API can't be used as `append` because:

- Users can't append on an exisiting path.
- Data is not readable during multipart upload.

Reply:
No, there is no `append` in `write`.

Reply:
> No, there is no `append` in `write`.

But the user could write multiple times on the same file. If someone do not understand the difference between multipart and append API, he may be confused. 

> Maybe need to update with `main` branch?

I have rebased onto the main branch. The failed ci was caused by a network issue...

<img width="576" alt="image" src="https://user-images.githubusercontent.com/32436597/234847763-08568aeb-08ec-41e8-9470-3bac2b89c907.png">

Reply:
> But the user could write multiple times on the same file. If someone do not understand 
the difference between multipart and append API, he may be confused.

Users won't know the different behavior between underlying services' `multipart` and `append`. They only know about the API provided by OpenDAL: `write` and `append`.

They have different sementics:

- `write`: Always create a new file, not readable unitl `close`.
- `append`: Can append existing file, readable after `append` return.

---

I have to make it clear that: please don't mix opendal's public API and underlying services API. Services can pick up different APIs to implement opendal's API. But our design needs to make it sure that users won't affected by the implement details.

Reply:
Ok, the proposal's goal will be to add the append API while leaving the write API untouched.